### PR TITLE
[MCM 2/3] How-to: Dynamic circuits with mid-circuit measurements

### PR DESCRIPTION
**Pull request by dwierichs**
**Title:**
How to create dynamic circuits with mid-circuit measurements

**Summary:**
This how-to is a walkthrough of creating a `QNode` that contains mid-circuit measurements, logical processing of the measured values, and conditioning functions on these processed measurements.
As a warmup task, we implement a T-gadget, which showcases a single MCM without processing that conditions a single follow-up gate.  This serves the purpose to have a minimal example of a dynamic circuit, in addition to the more contrived `QNode` example.

**Relevant references:**

**Possible Drawbacks:**

**Related GitHub Issues:**

----
If you are writing a demonstration, please answer these questions to facilitate the marketing process.

* GOALS — Why are we working on this now?

  This how-to promotes mid-circuit measurement functionality in PennyLane, which has been extended in recent releases.
  In addition, this how-to is part of the broader goal of emphasizing MCMs and dynamic circuits in our contents.

* AUDIENCE — Who is this for?

  Typical how-to audience: Fundamental concepts of measurements and quantum computing are known, but the audience may not yet be familiar with PennyLane's syntax for MCMs and conditioning circuit parts on their outcomes, or they may want to consult this how-to as a reference of how to realize their dynamic circuit with MCMs.
Anyone really with the goal of coding up a circuit that uses MCM values and feedforward can be in the audience.

* KEYWORDS — What words should be included in the marketing post?

  - Dynamic circuit
  - Mid-circuit measurements
  - Feedforward control
  - [optional] Measurement value processing

* Which of the following types of documentation is most similar to your file? 
(more details [here](https://www.notion.so/xanaduai/Different-kinds-of-documentation-69200645fe59442991c71f9e7d8a77f8))
    
- [ ] Tutorial
- [ ] Demo
- [x] How-to

### CODE REVIEW

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.metadata.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{
+    "title": "How to create dynamic circuits with mid-circuit measurements",
+    "authors": [
+        {
+            "id": "david_wierichs"
+        }
+    ],
+    "dateOfPublication": "2024-04-29T00:00:00+00:00",
+    "dateOfLastModification": "2024-04-29T00:00:00+00:00",
```
Comment (dwierichs) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.metadata.json:9]: placeholder comment regarding publication/modification dates.

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+Measuring qubits in the middle of a quantum circuit execution can be useful in many ways.
+From hardware characterization, modeling and error mitigation, over physical
+phenomena like 
+`measurement-induced entanglement phase transitions <https://arxiv.org/abs/1808.05953>`_,
+which may be used to enhance circuit trainability <https://scipost.org/SciPostPhys.14.6.147>`_,
+to error correction,
+algorithmic improvements and even up to full
+computations encoded as measurements in 
+`measurement-based quantum computation (MBQC) <link.todo.com>`_
```
Comment (dwierichs) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:12]: which link should be added here?

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+######################################################################
+# The MCM is performed with :func:`~.pennylane.measure` and
+# the conditioned application of the ``S`` gate is realized with :func:`~.pennylane.cond`.
+#
+# In case you wondered, this circuit implements a so-called
+# `T-gadget <https://arxiv.org/abs/quant-ph/0002039>`_ and the ``fixed_state`` we prepared on
+# the second qubit is called a `"magic state"
+# <https://en.wikipedia.org/wiki/Magic_state_distillation#Magic_states>`_, but this will not
+# concern us here.
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:63]: @dwierichs I would either remove this, or move this into the introduction of this section. E.g.,

```rest
We start with a minimal dynamic circuit on two qubits, which implements
a `T-gadget <https://arxiv.org/abs/quant-ph/0002039>`_ to prepare
a :doc:`magic state <tutorial_magic_state_distillation>`.
```

I've added a suggestion above for this.
```suggestion
```

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+# ---------------------------------------
+#
+# We now turn to a more complex example of a dynamic circuit. We will build a circuit that
+# non-deterministically initializes half-filled computational basis states, i.e. basis states
+# with as many :math:`1`\ s as :math:`0`\ s.
+#
+# The procedure is as follows: Single-qubit rotations and a layer of :class:`~.pennylane.CNOT`
+# gates create an entangled state on the first three qubits. Afterwards, these qubits are
+# measured and for each measured :math:`0`, another qubit is excited from the :math:`|0\rangle`
+# state to the :math:`|1\rangle` state.
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:79]: bullet points (even if just 2) might help here as well :)

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+# Indeed, we created half-filled computational basis states, each with its own probability:
+#
+
+print("The probabilities for the bit strings are:")
+for key, val in counts.items():
+    print(f"    {key}: {val/shots*100:4.1f} %")
+
+######################################################################
+# Quiz question: Did we sample *all* possible half-filled basis states at
+# least once? You can find the answer at the end of this how-to.
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:144]: @dwierichs my recommendation here is not to post questions without answering them straightaway :) People don't like scrolling if they already have what they need!



```suggestion
# .. note::
# 
#   You may be wondering, did we sample *all* possible half-filled basis states at
#   least once?
#
#   We did not create all possible
#   half-filled states at least once. This is because our circuit forces each of the qubit pairs
#   ``(0, 3)``, ``(1, 4)`` and ``(2, 5)`` to be in the states :math:`|01\rangle` or
#   :math:`|10\rangle`. However, there are half-filled states that do not have this form, like
#   ``100110`` for example, which you will not find among the sampled states from our circuit.
```

Comment (josh146) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:144]: Having said that, now it _definitely_ looks like an aside and something that doesn't belong in a how-to!

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+#
+# After this minimal working example, we now construct a more complex circuit showcasing more
+# features of MCMs and dynamic circuits in PennyLane. We start with some short preparatory
+# definitions.
+#
+# State preparation with dynamic circuits
+# ---------------------------------------
+#
+# We now turn to a more complex example of a dynamic circuit. We will build a circuit that
+# non-deterministically initializes half-filled computational basis states, i.e. basis states
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:73]:
```suggestion
# non-deterministically initializes half-filled computational basis states, i.e., basis states
```

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+# the second qubit is called a `"magic state"
+# <https://en.wikipedia.org/wiki/Magic_state_distillation#Magic_states>`_, but this will not
+# concern us here.
+#
+# After this minimal working example, we now construct a more complex circuit showcasing more
+# features of MCMs and dynamic circuits in PennyLane. We start with some short preparatory
+# definitions.
+#
+# State preparation with dynamic circuits
+# ---------------------------------------
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:70]: @dwierichs the intro of this section implies that the title should be something along the lines of 'How to use MCMs to initialize half-filled computational basis states'.

Do you have a shorter suggestion that might work as a title?

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+for key, val in counts.items():
+    print(f"    {key}: {val/shots*100:4.1f} %")
+
+######################################################################
+# Quiz question: Did we sample *all* possible half-filled basis states at
+# least once? You can find the answer at the end of this how-to.
+#
+# Controlling dynamic circuits with postselection
+# -----------------------------------------------
+# We may select only some of these half-filled states by postselecting on measurement outcomes:
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:148]:
```suggestion
# We can select only some of these half-filled states by postselecting on measurement outcomes:
```

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+print("The probabilities for the bit strings are:")
+for key, val in counts.items():
+    print(f"    {key}: {val/shots*100:4.1f} %")
+
+######################################################################
+# Quiz question: Did we sample *all* possible half-filled basis states at
+# least once? You can find the answer at the end of this how-to.
+#
+# Controlling dynamic circuits with postselection
+# -----------------------------------------------
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:147]: Should the section title be related to post-selecting states?

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+print("The probabilities for the postselected bit strings are:")
+for key, val in counts.items():
+    print(f"    {key}: {val/postselected_shots*100:4.1f} %")
+
+######################################################################
+# We successfully postselected on the desired properties of the computational basis state. Note
+# that the number of returned samples is reduced, because those samples that do not meet the
+# postselection criteria are discarded entirely.
+#
+# The quiz question from above may have become a bit easier to answer with this result...
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:192]:
```suggestion
```

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+# <https://docs.pennylane.ai/en/stable/introduction/measurements.html#mid-circuit-measurements-and-conditional-operations>`_
+# and the documentation of :func:`~.pennylane.measure`.
+#
+# And this is how to create dynamic circuits in PennyLane with mid-circuit measurements!
+#
+# Before finishing, here is the answer to the quiz question: We did not create all possible
+# half-filled states at least once. This is because our circuit forces each of the qubit pairs
+# ``(0, 3)``, ``(1, 4)`` and ``(2, 5)`` to be in the states :math:`|01\rangle` or
+# :math:`|10\rangle`. However, there are half-filled states that do not have this form, like
+# ``100110`` for example, which you will not find among the sampled states from our circuit.
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:256]:
```suggestion
```

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+# this evened out the probabilities of measuring either of the two possible bit strings.
+# Also, note that we conditionally
+# applied the bit flip operators ``qml.X`` by comparing an MCM result to
+# the corresponding selection criterion (``mcm!=sel``). More generally, MCM
+# results in PennyLane can be processed with standard arithmetic operations. For details,
+# see the `introduction to MCMs
+# <https://docs.pennylane.ai/en/stable/introduction/measurements.html#mid-circuit-measurements-and-conditional-operations>`_
+# and the documentation of :func:`~.pennylane.measure`.
+#
+# And this is how to create dynamic circuits in PennyLane with mid-circuit measurements!
```
Comment (josh146) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:250]: nice!

**Review by josh146 — APPROVED** [review]
Hey @dwierichs, looking really good! Approved but conditional on the comments :)

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
+r"""How to create dynamic circuits with mid-circuit measurements
+================================================================
+
+Measuring qubits in the middle of a quantum circuit execution can be useful in many ways.
+From understanding the inner workings of a circuit, hardware characterization,
+modeling and error mitigation, to error correction, algorithmic improvements and even up to full
+computations encoded as measurements in measurement-based quantum computation (MBQC).
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:7]: Good call! Added locally.

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+
+Before turning to any of these advanced topics, it is worthwhile to familiarize ourselves with
+the syntax and features around mid-circuit measurements (MCMs). In this how-to, we will focus on
+dynamic quantum circuits that use control flow based on MCMs.
+Most of the advanced concepts mentioned above incorporate MCMs in this way, making it a
+key ingredient to scalable quantum computing.
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:13]: Added a new sentence with these two scenarios in mind. Good? :)

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+
+.. figure:: ../_static/demonstration_assets/how_to_create_dynamic_mcm_circuits/socialthumbnail_how_to_create_dynamic_mcm_circuits.png
+    :align: center
+    :width: 50%
+
+"""
+
+######################################################################
+# Minimal working example
+# -----------------------
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:23]: How about "Dynamic circuit with a single MCM and conditional"? Even though it is not particularly short, I feel like it grasps the compactness of the section.

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+#
+# We start with a minimal dynamic circuit on two qubits. It rotates one qubit
+# about the ``X``-axis and prepares the other qubit in a fixed state.
+# After an entangling :class:`~.pennylane.CNOT` gate, the second qubit is measured,
+# and if we measured a ``1``, an :class:`~.pennylane.S` gate is applied.
+# Finally, the expectation value of the Pauli ``Y`` operator on the first qubit is returned.
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:29]: I like it a lot!

Are you set on introducing the T-gadget and magic state right here? I thought it might be distracting to drop those concepts before showing the cute little code example :thinking:

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 3 above)
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:63]: Haha only seeing this now. I think I'll just remove it!

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 7 above)
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:70]: You mean change the title of the how-to? :upside_down_face: 
The example is just the vessel to show a dynamic circuit :thinking: Should it be in the focus of the title?

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 5 above)
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:144]: Fair enough, removed it. Note, though that the MCM stats how-to _does_ contain such a quiz question with answer at the end :grimacing:

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 9 above)
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:147]: How about "Postselecting dynamically prepared states"? :)

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 7 above)
Comment (josh146) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:74]: Oh I mean, changing the section title here :) not the how to title

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (josh146) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py]: nice! I like it

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 5 above)
Comment (josh146) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py]: yeah I remember 😆

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 9 above)
Comment (josh146) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py]: nice!

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+
+dev = qml.device("lightning.qubit", wires=2)
+
+fixed_state = np.array([1.0, np.exp(1j * np.pi / 4)]) / np.sqrt(2)
+
+
+@qml.qnode(dev, interface="numpy")
+def circuit(x):
+    qml.RX(x, 0)
+
```
Comment (KetpuntoG) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:55]:
```suggestion
```

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+Before turning to any of these advanced topics, it is worthwhile to familiarize ourselves with
+the syntax and features around mid-circuit measurements (MCMs). In this how-to, we will focus on
+dynamic quantum circuits that use control flow based on MCMs.
+Most of the advanced concepts mentioned above incorporate MCMs in this way, making it a
+key ingredient to scalable quantum computing.
+
+If you want to learn more or are looking for how to collect statistics of mid-circuit measurements,
+have a read of the :doc:`how-to on MCM statistics </demos/tutorial_how_to_collect_mcm_stats>`.
+
```
Comment (KetpuntoG) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:18]: I think this how-to is necessary before going to the other one. Maybe you could put this sentence at the end of the how-to for those who want to continue learning.

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+# Before running this ``QNode``, let's sample some random input parameters and draw the circuit:
+#
+
+np.random.seed(652)
+x = np.random.random(3) * np.pi
+
+print(qml.draw(create_half_filled_state)(x))
+
+######################################################################
+# We see the initial state preparation and the measurement & conditional bit flip
```
Comment (KetpuntoG) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:132]:
```suggestion
# We see the initial state preparation and the measurement & conditional bit flip
```

Could you rephrase this sentence? the "&" is a bit confused to me

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+# :math:`1` in the third position. We do not postselect on the second qubit, which we can indicate
+# by passing ``None`` to the ``postselect`` argument of :func:`~.pennylane.measure`. Again, before
+# running the circuit, let's draw it first:
+#
+
+selection = [0, None, 1]
+print(qml.draw(postselect_half_filled_state)(x, selection))
+
+######################################################################
+# Note the indicated postselection values next to the drawn MCMs.
```
Comment (KetpuntoG) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:179]: cool!

**Review by KetpuntoG — COMMENTED** [review]
Easy and simple! A couple of clarifications and we will ready to merge :)

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 7 above)
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:74]: ah, sorry, misread this!
How about "How to dynamically prepare half-filled basis states"? :)

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 19 above)
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:132]: How about this? :) 

>We see an initial block of gates that prepares the starting state on the                
first three qubits, followed by pairs of measurements and conditional bit flips,        
applied to pairs of qubits.

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 18 above)
Comment (KetpuntoG) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:18]: Okey, sounds good

**Review by KetpuntoG — APPROVED** [review]
Great job!

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+# -------------------------------------------------
+#
+# We start with a minimal dynamic circuit on two qubits.
+#
+# - It rotates one qubit about the ``X``-axis and prepares
+#   the other qubit in a fixed state.
+#
+# - Both qubits are entangled with a :class:`~.pennylane.CNOT` gate.
+#
+# - The second qubit is measured with :class:`~.pennylane.measure`.
```
Comment (trbromley) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:36]:
```suggestion
# - The second qubit is measured with :func:`~.pennylane.measure`.
```

**Comment by KetpuntoG** [comment]
(Even if It doesn't use new features, I will target `dev` to ensure that there are no unexpected problems on Monday)

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+#
+# - Finally, the expectation value of the Pauli ``Y`` operator on the first qubit is returned.
+#
+
+import pennylane as qml
+import numpy as np
+
+dev = qml.device("lightning.qubit", wires=2)
+
+fixed_state = np.array([1.0, np.exp(1j * np.pi / 4)]) / np.sqrt(2)
```
Comment (trbromley) [inline demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:49]: Do we need the `fixed_state`?

**Review by trbromley — COMMENTED** [review]
Thanks @dwierichs! Just leaving two quick comments but nothing to block doing a merge.

File: demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py (hunk 22 above)
Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:49]: You mean to define it outside of the QNode?

Comment (KetpuntoG) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:49]: Maybe the question is whether you can directly generate it with a gate instead of specifying it and use the embedding

Comment (dwierichs) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:49]: Ah, I see. If that's the question, this is motivated by this being a T-gadget, where implementing the gate (`qml.T`) that would get us the state kind of is the goal. :D But as we are not mentioning this background (anymore), I'm not sure how important this is...

Comment (KetpuntoG) [reply demonstrations/tutorial_how_to_create_dynamic_mcm_circuits.py:49]: Could you put then a T gate? 
if that is the case that would be better

**Comment by KetpuntoG** [comment]
Thanks for that last change @dwierichs ! Once It pass the CI, if there is not issues,  I will do the merge. Is it ok for you? or anything else I should keep in mind?

**Comment by KetpuntoG** [comment]
> Thanks for that last change @dwierichs ! Once It pass the CI, if there is not issues, I will do the merge. Is it ok for you? or anything else I should keep in mind?

Oh I see the auto-merge, I'll consider it as a yes 😄
